clang/lib/AST/ItaniumMangle.cpp

```diff
@@ -2445,6 +2445,46 @@ static bool isTypeSubstitutable(Qualifiers Quals, const Type *Ty,
   return true;
 }
 
+namespace {
+using DeclContextDesc = std::pair<clang::Decl::Kind, StringRef>;
+bool matchQualifiedTypeName(const QualType &Ty,
+                            ArrayRef<DeclContextDesc> Scopes) {
+  // The idea: check the declaration context chain starting from the type
+  // itself. At each step check the context is of expected kind
+  // (namespace) and name.
+  const CXXRecordDecl *RecTy = Ty->getAsCXXRecordDecl();
+
+  if (!RecTy)
+    return false; // only classes/structs supported
+  const auto *Ctx = dyn_cast<DeclContext>(RecTy);
+  StringRef Name = "";
+
+  for (const auto &Scope : llvm::reverse(Scopes)) {
+    clang::Decl::Kind DK = Ctx->getDeclKind();
+
+    if (DK != Scope.first)
+      return false;
+
+    switch (DK) {
+    case clang::Decl::Kind::ClassTemplateSpecialization:
+      // ClassTemplateSpecializationDecl inherits from CXXRecordDecl
+    case clang::Decl::Kind::CXXRecord:
+      Name = cast<CXXRecordDecl>(Ctx)->getName();
+      break;
+    case clang::Decl::Kind::Namespace:
+      Name = cast<NamespaceDecl>(Ctx)->getName();
+      break;
+    default:
+      llvm_unreachable("matchQualifiedTypeName: decl kind not supported");
```

Comment:
Should probably handle linkage specs as well.  I don't think this should be unreachable, since we need to do this for ALL types.  This should just give up and return false I think.

Reply:
> Should probably handle linkage specs as well

@erichkeane, I don't really understand what do you mean by "linkage spec". Do I need to check for linkage attributes of the type? Or you mean that there are several more kinds of declarations to be checked here?

Reply:
The latter.  This ends up hitting 'llvm_unreachable' when we might be testing user types here.  For example, what if we tested:

namespace foo {
extern "C++" class half;
}

I'd expect this to return 'false', not crash.

Reply:
The example you mentioned doesn't cause compiler crashes, so I will mark this as done